[SecretsManager] Fixes, Enhancements, Snapshot tests

MEPalma opened the pull request:
Migrated test suite to use snapshots with the exception of secret rotation tests and duplicate http tests.
Applied a number of fixes and enhancements.

Enhancements/Fixes/Patches:
- correct handling of given secret ids in the form of ARNs with trailing '-'
- update of secret's LastAccessedDate field on GetSecretValue requests
- added support for LastAccessedDate about secret versions
- update of secret's version LastAccessedDate fields
- fixed output format of ListSecretVersionIds request that now reports the exact CreatedDate instead of current time
- fixed output format of ListSecretVersionIds request that now reports the version's LastAccessedDate if available
- fixed output format of ListSecretVersionIds request that now returns results in the correct order
- enhanced PutSecretValue requests to mimic AWS's behaviour in its insertion order where this was consistent (with AWS* tags, otherwise sporadic)
- fixed UpdateSecret request handling that now creates a new secret version iff changes to relevant details are provided (SecretString, SecretBinary KmsKeyId), otherwise the current version is updated (eg. description)
- enhanced UpdateSecretVersionStage request handling that now versions stages are added in front not appended, AWSCURRENT is always in front of all other AWS stages (as this was the most consistent behaviour displayed by AWS, otherwise inconsistent ordering)
- fixed `SecretNotFoundException` sometimes returning an incorrect error code

Added relevant snapshot support, such as:
- DropUnsupportedEntries: which drops all unsupported fields from the snapshot
- NormaliseVersionStages which ensures validation about VersionStages field is performed against a constant sorting strategy (AWS displayed inconsistent sorting habits about this field)
- Others such as version_uuid pattern matching

dominikschubert reviewed (changes requested):
Love the extended test coverage :+1: 

Added some comments where I saw that the code style or general usage diverges too far from the current standard or where we already have specific patterns in place.

One issue I've noticed is that the resource cleanup needs to be adjusted to be more robust. A typical example for a resource leak here would be:
```
# 1. create secret
# 2. do something with secret
# 3. cleanup (delete secret)
```

If anything in step 2 fails, the cleanup will be skipped and resources start leaking. This affects LocalStack as well but more importantly your AWS account might end up cluttered with these.

There are a few patterns for this in place right now:

1. Create a fixture
2. Using the `cleanups` fixture and appending cleanup callables (in the simplest case just a lambda performing the same cleanup you'd otherwise do)
3. try/finally blocks (these additionally need to account for the fact that when cleaning up multiple resources, subsequent resource cleanup has to stay unaffected)

Option 1/2 are in general preferred over 3

MEPalma commented on tests/integration/secretsmanager/test_secretsmanager.py:
```diff
@@ -45,76 +58,141 @@ class TestSecretsManager:
+            )
+            secret_ids: Set[str] = {secret["Name"] for secret in lst.get("SecretList", [])}
+            return secret_id in secret_ids
+
+        success = poll_condition(condition=_is_secret_in_list, timeout=20, interval=2)
+        if not success:
+            LOG.warning(f"Timed out whilst awaiting for secret '{secret_id}' to become listable.")
+
+    @staticmethod
+    def _wait_force_deletion_completed(client, secret_id: str):
+        def _is_secret_deleted():
+            deleted = False
+            try:
+                client.describe_secret(SecretId=secret_id)
+            except Exception as ex:
+                if ex.response["Error"]["Code"] == "ResourceNotFoundException":
+                    deleted = True
+                else:
+                    raise ex
+            return deleted
+
+        success = poll_condition(condition=_is_secret_deleted, timeout=120, interval=30)
+        if not success:
+            LOG.warning(
+                f"Timed out whilst awaiting for force deletion of secret '{secret_id}' to complete."
+            )
+
+    @staticmethod
+    def _normalise_list_secret_version_ids_response(response: ListSecretVersionIdsResponse) -> None:
+        versions: List[SecretVersionsListEntry] = response["Versions"]
+        versions.sort(key=lambda e: e["CreatedDate"], reverse=True)
+
     @pytest.mark.parametrize(
         "secret_name, is_valid_partial_arn",
         [
-            (f"s-{short_uid()}", True),
+            ("s-c64bdc03", True),
             ("Valid/_+=.@-Name", True),
             ("Valid/_+=.@-Name-a1b2", True),
+            ("Valid/_+=.@-Name-a1b2c3-", True),
```
Yes no duplicates are allowed, however I don't see how snapshots are gonna reference this test if the parameter keeps changing? Although static the probability of this being equal to some other randomly generated (but not static) secret name I believe is negligible, no?

MEPalma commented on localstack/services/secretsmanager/provider.py:
```diff
@@ -358,6 +364,60 @@ def moto_smb_create_secret(fn, self, name, *args, **kwargs):
+def moto_smb_put_secret_value(
+    fn, self, secret_id, secret_string, secret_binary, client_request_token, version_stages
+):
+    # Patch: VersionStages are sequentially appended but AWS* ones which are appended after custom stages.
+    # TODO: more investigations are required, AWS behaviour seems sporadic.
```
Yes I agree actually, I'd say we should consider randomising this then, as implementers might assume AWS itself being consistent when it's not (as far as i can tell!). I have removed this for now as you suggested.

MEPalma commented on localstack/services/secretsmanager/provider.py:
```diff
@@ -358,6 +364,60 @@ def moto_smb_create_secret(fn, self, name, *args, **kwargs):
+@patch(SecretsManagerBackend.list_secret_version_ids)
+def moto_smb_list_secret_version_ids(_, self, secret_id, *args, **kwargs):
+    if secret_id not in self.secrets:
+        raise SecretNotFoundException()
```
Yes indeed, the exception's message was missing, and the error type was wrong. thanks for spotting this 👍

MEPalma commented:
Thank you @dominikschubert for this review 🙏 . It sure helped me figure out some Snapshots anti-patterns I had used in this first exercise for me. Hopefully this latest revision is a much better candidate.

dominikschubert commented on localstack/testing/snapshots/transformer_utility.py:
```diff
@@ -317,6 +323,72 @@ def cloudwatch_api():
 
+    @staticmethod
+    def secretsmanager_api():
+        class NormaliseVersionStages(Transformer):
+            """
+            Normalises Version Stages list in responses so that these are sorted. This is to enable a set comparison of
+            the Version Stages in all responses, as AWS's ordering behaviour is unknown.
+            """
+
+            def _normalise(self, input_data: dict | list) -> dict | list:
+                if isinstance(input_data, dict):
+                    res = dict()
+                    for k, v in input_data.items():
+                        res[k] = (
+                            sorted(v)
+                            if k == "VersionStages" and isinstance(v, list)
+                            else self._normalise(v)
+                        )
+                    return res
+                elif isinstance(input_data, list):
+                    return list(map(self._normalise, input_data))
+                else:
+                    return input_data
+
+            def transform(self, input_data: dict, *, ctx: TransformContext) -> dict:
+                return self._normalise(input_data)
+
+        class NormaliseSecretVersions(Transformer):
+            """
+            Normalises ListSecretVersionIdsResponse responses so that Secret Versions are sorted by date. This is to
+            enable a set comparison of the Secret Versions in the response, as AWS's ordering behaviour is unknown.
+            """
+
+            def transform(self, input_data: dict, *, ctx: TransformContext) -> dict:
+                for k, v in input_data.items():
+                    if k == "Versions" and isinstance(v, list):
+                        input_data[k] = sorted(v, key=lambda e: e.get("CreatedDate"), reverse=True)
+                    elif isinstance(v, dict):
+                        input_data[k] = self.transform(v, ctx=ctx)
+                return input_data
```
might be able to use the new `SortingTransformer` for this since it covers more or less exactly this purpose. Especially for `NormaliseSecretVersions` this would be more or less a direct mapping to something like `SortingTransformer("Versions", lambda e: e.get("CreatedDate"))`

dominikschubert commented on tests/integration/secretsmanager/test_secretsmanager.py:
```diff
@@ -335,147 +448,143 @@ def test_rotate_secret_invalid_lambda_arn(self, sm_client):
         with pytest.raises(Exception) as validation_exception:
             sm_client.create_secret(Name=secret_name, SecretString="MySecretString")
-        check_validation_exception(validation_exception)
+        sm_snapshot.match("ex_log_1", validation_exception.value)
```
```suggestion
        sm_snapshot.match("ex_log_1", validation_exception.value.response)
```

Just saw that I forgot a crucial part of this when suggesting it previously

It should actually be `<exception>.value.response` instead of `<exception>.value`
(Same for the ones below)

dominikschubert reviewed (changes requested):
Minor suggestion re: transformer.

Also added a suggestion for capturing the error response with a better shape. When that one at least is addressed it'll be ready to merge :rocket:

MEPalma commented:
Thank you @dominikschubert 
Latest changes include:
- fix `SecretNotFoundException`'s error code (see patch)
- updated `SortingTransformer` to search in lists
- swapped the secretsmanager's normalising transformers for the `SortingTransformer` where appropriate, as suggested
- fix exception's snapshots object as suggested
- updated snapshots

dominikschubert reviewed (approved):
LGTM now! Thanks for addressing all the comments and suggestions! :rocket:
